triton-cli/src/main.rs

```diff
@@ -0,0 +1,259 @@
+};
+use triton_vm::prelude::*;
+
+#[derive(Debug, Parser)] // requires `derive` feature
+#[command(name = "triton-cli")]
+#[command(about = "Compile, prove and verify tasm programs", long_about = None)]
+struct Cli {
+    #[command(subcommand)]
+    command: Commands,
+}
+
+#[derive(Debug, Subcommand)]
+enum Commands {
+    #[command(arg_required_else_help = true)]
+    Prove {
+        asm_path: String,
+        proof_out_path: String,
+        #[clap(long)]
+        /// Public inputs to pass to the proof, comma separated
+        public_inputs: Option<String>,
+        #[clap(long)]
+        /// Private inputs to pass to the proof, comma separated
+        private_inputs: Option<String>,
+    },
+    #[command(arg_required_else_help = true)]
+    Verify { proof_path: String },
+}
+
+fn main() {
+    let args = Cli::parse();
+    match args.command {
+        Commands::Prove {
+            asm_path,
+            proof_out_path,
+            public_inputs,
+            private_inputs,
+        } => prove(&asm_path, &proof_out_path, public_inputs, private_inputs),
+        Commands::Verify { proof_path } => verify(&proof_path),
+    }
+}
+
+fn digest_to_str(d: Digest) -> String {
+    let mut hex = String::new();
+    d.0.iter()
+        .for_each(|v| write!(&mut hex, "{:16x}", u64::from(v)).unwrap());
+    format!("0x{hex}")
+}
+
+fn verify(proof_path: &String) {
+    let (stark, claim, proof) = read_proof(proof_path).expect("Failed to load proof");
+
+    let verdict = triton_vm::verify(stark, &claim, &proof);
+    if !verdict {
+        println!("Proof is not valid!");
+        return;
+    }
+    println!("proof is valid!");
+    println!("program digest: {}", digest_to_str(claim.program_digest));
+    println!("=================");
+    println!("security level: {} bits", stark.security_level);
+    println!("FRI expansion factor: {}", stark.fri_expansion_factor);
+    println!("trace randomizers: {}", stark.num_trace_randomizers);
+    println!("colinearity checks: {}", stark.num_collinearity_checks);
+    println!("codeword checks: {}", stark.num_combination_codeword_checks);
+    println!("=================");
+    println!("public inputs:");
+    if claim.input.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.input {
+        println!("{}", u64::from(v));
+    }
+    println!("public outputs:");
+    if claim.output.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.output {
+        println!("{}", u64::from(v));
+    }
+}
+
+fn parse_inputs(inputs: Option<String>) -> Vec<BFieldElement> {
+    inputs
+        .unwrap_or_default()
+        .split(',')
+        .filter(|v| !v.is_empty())
+        .map(|v| BFieldElement::from(v.parse::<u64>().unwrap()))
+        .collect::<Vec<BFieldElement>>()
+}
+
+fn prove(
+    asm: &String,
+    out: &String,
+    public_inputs: Option<String>,
+    private_inputs: Option<String>,
+) {
+    // TODO: test paths before making the proof?
```

Comment:
Also, silently overwriting an existing file might not make everybody happy. It could be a possibility to abort & warn the user if the output file already exists. I'm not sure where I stand on this, though; other CLI utilities silently overwrite existing files also, so there certainly is precedence.

Reply:
It's the opposite, if the file exists it throws an error. But this is almost worse because if a large proof is built it will be discarded if the output filepath already exists.